src/vm/values/types.nim

```diff
@@ -234,28 +240,67 @@ type
 {.hint: "'Value's inner type is currently '" & $sizeof(ValueObj) & "'.".}
 {.hints: off.}
 
-when sizeof(ValueObj) > 64: # At time of writing it was '56', 8 - 64 bit integers seems like a good warning site? Can always go smaller
+when sizeof(ValueObj) > 72: # At time of writing it was '72', 8 - 64 bit integers seems like a good warning site? Can always go smaller
     {.warning: "'Value's inner object is large which will impact performance".}
 
-template makeFuncAccessor*(name: untyped) =
-  template name*(val: Value): typeof(val.funcType.name) =
-    assert val.kind == Function
-    val.funcType.name
-
-
-makeFuncAccessor(args)
-makeFuncAccessor(attrs)
-makeFuncAccessor(arity)
-makeFuncAccessor(returns)
-makeFuncAccessor(example)
-makeFuncAccessor(fnKind)
-makeFuncAccessor(params)
-makeFuncAccessor(main)
-makeFuncAccessor(imports)
-makeFuncAccessor(exports)
-makeFuncAccessor(memoize)
-makeFuncAccessor(bcode)
-makeFuncAccessor(inline)
-makeFuncAccessor(action)
+template readonly*(val: Value): bool = isReadOnly in val.flags
+template `readonly=`*(val: Value, newVal: bool) = val.flags[isReadOnly] = newVal
+
+template dirty*(val: Value): bool = isDirty in val.flags
+template `dirty=`*(val: Value, newVal: bool) = val.flags[isDirty] = newVal
+
+template dynamic*(val: Value): bool = isDynamic in val.flags
+template `dynamic=`*(val: Value, newVal: bool) = val.flags[isDynamic] = newVal
+
+proc b*(val: Value): VLogical {.inline.} =
+    assert (val.flags * {isFalse, isMaybe, isTrue}).len == 1
+    if isMaybe in val.flags:
+        Maybe
+    elif isTrue in val.flags:
+        True
+    else:
+        False
+
+template `b=`*(val: Value, newVal: VLogical) =
+    assert val.kind == Logical
+    val.flags.excl {isTrue, isFalse, isMaybe}
+    const flags = [False: isFalse, True: isTrue, Maybe: isMaybe]
+    val.flags.incl flags[newVal]
+
+template isFalse*(val: Value): bool = isFalse in val.flags
+template isMaybe*(val: Value): bool = isMaybe in val.flags
+template isTrue*(val: Value): bool = isTrue in val.flags
+
+
+template makeAccessor(field, subfield: untyped) =
+    template subfield*(val: Value): typeof(val.field.subfield) =
+        let myVal = val.field.subfield
+        myVal
```

Comment:
The changes are all looking good. Good job!

I have only one issue:

Is there some specific reason this ^ is done this way?

I mean... I generally tend to avoid every possible `let` or `var` (since it usually leads to unnecessary copying).

In that case, given that these fields are used quite a lot (e.g. on function calls) and some of the fields are arrays (`seq`s), I'm looking into the generated C code and I'm seeing quite a few `eqcopy`s for this `myVal`.

Couldn't we just "alias" subfield with `val.field.subfield`? Is there some issue with that approach?

Reply:
Whoops that was me fighting around the issue when `b` is a template.

Reply:
`.b` that is trying to get a Logical value's value... doesn't (or won't better said) won't matter too much.

The way you've cleaned up the code has given me a great idea:

As I've said, in many many instances, I check for things like: `if x.kind==Logical and x.b == True`, etc. Given the way `Logical` values are defined right now, this whole construct will be transformed to a simple `isTrue(x)`. And obviously, we won't really have to bother what `.b` is... 😉